diff --git a/dist/adapters/cloudflare.mjs b/dist/adapters/cloudflare.mjs
index 402eb25cbe1a5f23af06a89ec2b78579278f1c7a..749043d7a9915949508ecfd851365b718d829d53 100644
--- a/dist/adapters/cloudflare.mjs
+++ b/dist/adapters/cloudflare.mjs
@@ -74,12 +74,13 @@ const cloudflareAdapter = (opts = {}) => {
         headers: upgradeHeaders
       });
     },
-    handleDurableInit: async (obj, state, env) => {
+    handleDurableInit: (obj, state, env) => {
+      state.blockConcurrencyWhile(async () => {
+        obj.context = await state.storage.get("__context") || {};
+      });
     },
     handleDurableUpgrade: async (obj, request) => {
-      const { upgradeHeaders, endResponse, namespace } = await hooks.upgrade(
-        request
-      );
+      const { upgradeHeaders, endResponse, context, namespace } = await hooks.upgrade(request);
       if (endResponse) {
         return endResponse;
       }
@@ -96,6 +97,8 @@ const cloudflareAdapter = (opts = {}) => {
       peers.add(peer);
       obj.ctx.acceptWebSocket(server);
       await hooks.callHook("open", peer);
+      await obj.ctx.storage.put("__context", context);
+      obj.context = context;
       return new Response(null, {
         status: 101,
         webSocket: client,
@@ -112,6 +115,8 @@ const cloudflareAdapter = (opts = {}) => {
       peers.delete(peer);
       const details = { code, reason, wasClean };
       await hooks.callHook("close", peer, details);
+      await obj.ctx.storage.deleteAlarm();
+      await obj.ctx.storage.deleteAll();
     },
     handleDurablePublish: async (_obj, topic, data, opts2) => {
       return durablePublish(topic, data, opts2);
@@ -181,6 +186,7 @@ class CloudflareDurablePeer extends Peer {
     peer = ws._crosswsPeer = new CloudflareDurablePeer({
       ws,
       request: request || new StubRequest(state.u || ""),
+      context: durable.context,
       namespace: namespace || state.n || "",
       durable
     });
