diff --git a/nodejs/loro_wasm.js b/nodejs/loro_wasm.js
index 3d22132c9b694a7b377ac06bd13cc80dcc86fbc2..d08c2edb19953c9e4d465f6f080f65b9cc09f108 100644
--- a/nodejs/loro_wasm.js
+++ b/nodejs/loro_wasm.js
@@ -7088,7 +7088,7 @@ module.exports.__wbindgen_typeof = function(arg0) {
     return addHeapObject(ret);
 };
 
-const path = require('path').join(__dirname, 'loro_wasm_bg.wasm');
+const path = new URL('./loro_wasm_bg.wasm', import.meta.url);
 const bytes = require('fs').readFileSync(path);
 
 const wasmModule = new WebAssembly.Module(bytes);
@@ -7101,7 +7101,7 @@ wasm.__wbindgen_start();
 
 // Don't patch this if it already exists (for example in Deno)
 if (!globalThis.crypto) {
-    // We need this patch because we use `getrandom` crate in Rust, which relies on this patch 
+    // We need this patch because we use `getrandom` crate in Rust, which relies on this patch
     // for nodejs
     // https://docs.rs/getrandom/latest/getrandom/#nodejs-es-module-support
     const { webcrypto } = require("crypto");
