diff --git a/nodejs/loro_wasm.js b/nodejs/loro_wasm.js
index 3d22132c9b694a7b377ac06bd13cc80dcc86fbc2..d944b95833f0d12cb28c0172348c502f71e1040b 100644
--- a/nodejs/loro_wasm.js
+++ b/nodejs/loro_wasm.js
@@ -7088,10 +7088,9 @@ module.exports.__wbindgen_typeof = function(arg0) {
     return addHeapObject(ret);
 };
 
-const path = require('path').join(__dirname, 'loro_wasm_bg.wasm');
-const bytes = require('fs').readFileSync(path);
-
-const wasmModule = new WebAssembly.Module(bytes);
+const wasmModule = process.env.NODE_ENV === "development"
+    ? new WebAssembly.Module(require('fs').readFileSync(require('path').join(__dirname, 'loro_wasm_bg.wasm')))
+    : require("./loro_wasm_bg.wasm?module");
 const wasmInstance = new WebAssembly.Instance(wasmModule, imports);
 wasm = wasmInstance.exports;
 module.exports.__wasm = wasm;
