diff --git a/dist/awareness.js b/dist/awareness.js
index 36817c3261c6717a649c5b361f593d34c4748dc5..de0be74d9f5ae2bc2cc4a89fd81518046cce78a7 100644
--- a/dist/awareness.js
+++ b/dist/awareness.js
@@ -204,27 +204,31 @@ export class AwarenessPlugin {
         this.getTextFromDoc = getTextFromDoc;
         this.sub = this.doc.subscribe((e) => {
             if (e.by === "local") {
+                Promise.resolve().then(() => {
                 // update remote cursor position
-                const effects = [];
-                for (const peer of this.awareness.peers()) {
-                    const effect = getEffects(this.doc, this.awareness, peer);
-                    if (effect) {
-                        effects.push(effect);
+                    const effects = [];
+                    for (const peer of this.awareness.peers()) {
+                        const effect = getEffects(this.doc, this.awareness, peer);
+                        if (effect) {
+                            effects.push(effect);
+                        }
                     }
-                }
-                this.view.dispatch({
-                    effects,
+                    this.view.dispatch({
+                        effects,
+                    });
                 });
             }
             else if (e.by === "checkout") {
-                // TODO: better way
-                this.view.dispatch({
-                    effects: [
-                        remoteAwarenessEffect.of({
-                            type: "checkout",
-                            checkout: this.doc.isDetached(),
-                        }),
-                    ],
+                Promise.resolve().then(() => {
+                    // TODO: better way
+                    this.view.dispatch({
+                        effects: [
+                            remoteAwarenessEffect.of({
+                                type: "checkout",
+                                checkout: this.doc.isDetached(),
+                            }),
+                        ],
+                    });
                 });
             }
         });
diff --git a/dist/ephemeral.js b/dist/ephemeral.js
index 543e8631d27513da10ab06cbbbc6e5c93862fa78..82f64bf56fa328ba62c040ab5801cd00adeb1a6b 100644
--- a/dist/ephemeral.js
+++ b/dist/ephemeral.js
@@ -99,31 +99,33 @@ export class EphemeralPlugin {
         this.initUser = false;
         this.sub = this.doc.subscribe((e) => {
             if (e.by === "local") {
-                // update remote cursor position
-                const { remoteCursors: remoteStates, isCheckout } = view.state.field(ephemeralStateField);
-                if (isCheckout)
-                    return;
-                const effects = [];
-                for (const peer of remoteStates.keys()) {
-                    if (peer === this.doc.peerIdStr) {
-                        continue;
-                    }
-                    const state = this.ephemeralStore.get(`${peer}-cm-cursor`);
-                    if (state) {
-                        const effect = getCursorEffect(this.doc, peer, state);
-                        if (effect) {
-                            effects.push(effect);
+                Promise.resolve().then(() => {
+                    // update remote cursor position
+                    const { remoteCursors: remoteStates, isCheckout } = view.state.field(ephemeralStateField);
+                    if (isCheckout)
+                        return;
+                    const effects = [];
+                    for (const peer of remoteStates.keys()) {
+                        if (peer === this.doc.peerIdStr) {
+                            continue;
+                        }
+                        const state = this.ephemeralStore.get(`${peer}-cm-cursor`);
+                        if (state) {
+                            const effect = getCursorEffect(this.doc, peer, state);
+                            if (effect) {
+                                effects.push(effect);
+                            }
+                        }
+                        else {
+                            effects.push(ephemeralEffect.of({
+                                type: "delete",
+                                peer,
+                            }));
                         }
                     }
-                    else {
-                        effects.push(ephemeralEffect.of({
-                            type: "delete",
-                            peer,
-                        }));
-                    }
-                }
-                this.view.dispatch({
-                    effects,
+                    this.view.dispatch({
+                        effects,
+                    });
                 });
             }
             else if (e.by === "checkout") {
diff --git a/dist/sync.js b/dist/sync.js
index 8f36aa0b82011dbc2980b0bfdd140fc31ede82dc..c7589f0d5828a080f036b39921e7f9bfe77baaa5 100644
--- a/dist/sync.js
+++ b/dist/sync.js
@@ -14,54 +14,58 @@ export class LoroSyncPluginValue {
             }
             if (e.by === "checkout") {
                 // TODO: better handle checkout
-                this.view.dispatch({
-                    changes: [
-                        {
-                            from: 0,
-                            to: this.view.state.doc.length,
-                            insert: this.getTextFromDoc(this.doc).toString(),
-                        },
-                    ],
-                    annotations: [loroSyncAnnotation.of(this)],
-                });
+                Promise.resolve().then(() => {
+                    this.view.dispatch({
+                        changes: [
+                            {
+                                from: 0,
+                                to: this.view.state.doc.length,
+                                insert: this.getTextFromDoc(this.doc).toString(),
+                            },
+                        ],
+                        annotations: [loroSyncAnnotation.of(this)],
+                    });
+                })
                 return;
             }
             if (e.by === "import") {
-                let changes = [];
-                let pos = 0;
-                for (let { diff, target } of e.events) {
-                    const text = this.getTextFromDoc(this.doc);
-                    // Skip if the event is not a text event
-                    if (diff.type !== "text")
-                        return;
-                    // Skip if the event is not for the current document
-                    if (target !== text.id)
-                        return;
-                    const textDiff = diff.diff;
-                    for (const delta of textDiff) {
-                        if (delta.insert) {
-                            changes.push({
-                                from: pos,
-                                to: pos,
-                                insert: delta.insert,
-                            });
-                        }
-                        else if (delta.delete) {
-                            changes.push({
-                                from: pos,
-                                to: pos + delta.delete,
-                            });
-                            pos += delta.delete;
-                        }
-                        else if (delta.retain != null) {
-                            pos += delta.retain;
+                Promise.resolve().then(() => {
+                    let changes = [];
+                    let pos = 0;
+                    for (let { diff, target } of e.events) {
+                        const text = this.getTextFromDoc(this.doc);
+                        // Skip if the event is not a text event
+                        if (diff.type !== "text")
+                            return;
+                        // Skip if the event is not for the current document
+                        if (target !== text.id)
+                            return;
+                        const textDiff = diff.diff;
+                        for (const delta of textDiff) {
+                            if (delta.insert) {
+                                changes.push({
+                                    from: pos,
+                                    to: pos,
+                                    insert: delta.insert,
+                                });
+                            }
+                            else if (delta.delete) {
+                                changes.push({
+                                    from: pos,
+                                    to: pos + delta.delete,
+                                });
+                                pos += delta.delete;
+                            }
+                            else if (delta.retain != null) {
+                                pos += delta.retain;
+                            }
                         }
+                        this.view.dispatch({
+                            changes,
+                            annotations: [loroSyncAnnotation.of(this)],
+                        });
                     }
-                    this.view.dispatch({
-                        changes,
-                        annotations: [loroSyncAnnotation.of(this)],
-                    });
-                }
+                });
             }
         };
         this.sub = doc.subscribe(this.onRemoteUpdate);
diff --git a/dist/undo.js b/dist/undo.js
index e29854abbaaccc0457a06f9fa959b559e4b20918..229687b25d77b94b5bc937a7d177f4d838be4810 100644
--- a/dist/undo.js
+++ b/dist/undo.js
@@ -37,41 +37,43 @@ export class UndoPluginValue {
         this.sub = doc.subscribe((e) => {
             if (e.origin !== "undo")
                 return;
-            let changes = [];
-            let pos = 0;
-            for (let { diff, target } of e.events) {
-                const text = this.getTextFromDoc(this.doc);
-                // Skip if the event is not a text event
-                if (diff.type !== "text")
-                    return;
-                // Skip if the event is not for the current document
-                if (target !== text.id)
-                    return;
-                const textDiff = diff.diff;
-                for (const delta of textDiff) {
-                    if (delta.insert) {
-                        changes.push({
-                            from: pos,
-                            to: pos,
-                            insert: delta.insert,
-                        });
-                    }
-                    else if (delta.delete) {
-                        changes.push({
-                            from: pos,
-                            to: pos + delta.delete,
-                        });
-                        pos += delta.delete;
-                    }
-                    else if (delta.retain != null) {
-                        pos += delta.retain;
+            Promise.resolve().then(() => {
+                let changes = [];
+                let pos = 0;
+                for (let { diff, target } of e.events) {
+                    const text = this.getTextFromDoc(this.doc);
+                    // Skip if the event is not a text event
+                    if (diff.type !== "text")
+                        return;
+                    // Skip if the event is not for the current document
+                    if (target !== text.id)
+                        return;
+                    const textDiff = diff.diff;
+                    for (const delta of textDiff) {
+                        if (delta.insert) {
+                            changes.push({
+                                from: pos,
+                                to: pos,
+                                insert: delta.insert,
+                            });
+                        }
+                        else if (delta.delete) {
+                            changes.push({
+                                from: pos,
+                                to: pos + delta.delete,
+                            });
+                            pos += delta.delete;
+                        }
+                        else if (delta.retain != null) {
+                            pos += delta.retain;
+                        }
                     }
+                    this.view.dispatch({
+                        changes,
+                        annotations: [loroSyncAnnotation.of("undo")],
+                    });
                 }
-                this.view.dispatch({
-                    changes,
-                    annotations: [loroSyncAnnotation.of("undo")],
-                });
-            }
+            });
         });
         this.undoManager.setOnPop((isUndo, value, counterRange) => {
             var _a, _b;
