diff --git a/dist/presets/cloudflare/runtime/cloudflare-durable.mjs b/dist/presets/cloudflare/runtime/cloudflare-durable.mjs
index daaf8185b9a086dfd0eb0edf17768cce45392582..49d0e12e824f442d66a7a6d5c21c1b910562ed86 100644
--- a/dist/presets/cloudflare/runtime/cloudflare-durable.mjs
+++ b/dist/presets/cloudflare/runtime/cloudflare-durable.mjs
@@ -1,6 +1,6 @@
 import "#nitro-internal-pollyfills";
 import { DurableObject } from "cloudflare:workers";
-import wsAdapter from "crossws/adapters/cloudflare-durable";
+import wsAdapter from "crossws/adapters/cloudflare";
 import { useNitroApp } from "nitropack/runtime";
 import { isPublicAssetURL } from "#nitro-internal-virtual/public-assets";
 import { createHandler, fetchHandler } from "./_module-handler.mjs";
@@ -34,6 +34,8 @@ export default createHandler({
   }
 });
 export class $DurableObject extends DurableObject {
+  context;
+
   constructor(state, env) {
     super(state, env);
     state.waitUntil(
